source-oracle/source_oracle/resources.py

```diff
@@ -0,0 +1,158 @@
+                "autoextensible": autoextensible,
+                "flashback_on": flashback_on,
+                "avg_retention_seconds": avg_retention_seconds,
+            })
+
+            if not flashback_on:
+                raise Exception("Flashback must be enabled on the database. See go.estuary.dev/source-oracle for more information.")
+
+            if undo_retention_seconds < one_week_seconds:
+                msg = f"We require a minimum of 7 days UNDO_RETENTION to ensure consistency of this task. The current UNDO_RETENTION is {undo_retention_seconds} seconds. See go.estuary.dev/source-oracle for more information on how to configure the UNDO_RETENTION."  # nopep8
+                if skip_retention_checks:
+                    log.warn(msg)
+                else:
+                    raise Exception(msg)
+
+            if avg_retention_seconds < one_week_seconds:
+                msg = f"We require a minimum of 7 days undo retention to ensure consistency of this task. The current average auto-tuned retention of the database for the past four days is {avg_retention_seconds} seconds. See go.estuary.dev/source-oracle for more information on how to configure the UNDO_RETENTION."  # nopep8
+                if skip_retention_checks:
+                    log.warn(msg)
+                else:
+                    raise Exception(msg)
+
+            if not autoextensible:
+                log.warn("We recommend making your undo tablespace auto-extensible. See go.estuary.dev/source-oracle for more information.")
+
+            if retention_mode != 'GUARANTEE':
+                log.warn("We recommend guaranteeing retention of the undo tablespace. See go.estuary.dev/source-oracle for more information.")
+
+
+async def all_resources(
+    log: Logger, http: HTTPMixin, config: EndpointConfig, pool: oracledb.AsyncConnectionPool,
+) -> list[common.Resource]:
+    resources_list = []
+
+    oracle_tables = await fetch_tables(log, pool)
+    owners = set([t.owner for t in oracle_tables])
+    oracle_columns = await fetch_columns(log, pool, owners)
+
+    current_scn = None
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            await c.execute("SELECT current_scn FROM V$DATABASE")
+            current_scn = (await c.fetchone())[0]
+
+    for ot in oracle_tables:
+        columns = [col for col in oracle_columns if col.table_name == ot.table_name and col.owner == ot.owner]
+        t = build_table(config, ot.owner, ot.table_name, columns)
+
+        max_rowid = None
+        async with pool.acquire() as conn:
+            with conn.cursor() as c:
+                await c.execute(f"SELECT max(ROWID) FROM {t.owner}.{t.table_name}")
+                max_rowid = (await c.fetchone())[0]
+        # if max_rowid is None, that maens there are no rows in the table, so we
+        # skip backfill
+        backfill = ResourceState.Backfill(cutoff=(max_rowid,)) if max_rowid is not None else None
```

Comment:
I'm going to be a bit pedantic here, but is it actually correct to store the maximum RowID in the cutoff field here? According to the class docstrings the cutoff value is supposed to be "[The] LogCursor at which incremental replication began" while RowID values are page cursors.

It feels like either the docstring and type of the cutoff property ought to be modified, or else a new property should be added to `ResourceState.Backfill` to explicitly model the concept of a backfill only proceeding up to a particular page cursor.